torch/distributed/_sharded_tensor/api.py

```diff
@@ -336,110 +329,145 @@ def _init_rpc(self):
+                )
 
-            if local_shard_tensor.layout != tensor_properties.layout:
+            if local_shard_tensor.device != local_device:
                 raise ValueError(
-                    f'Local shard tensor layout does not match with tensor_properties! '
-                    f'local shard tensor layout: {local_shard_tensor.dtype}, '
-                    f'tensor_properties layout: {tensor_properties.layout}'
+                    f"Local shard tensor device does not match with local Shard's placement! "
+                    f"Found local shard tensor device: {local_shard_tensor.device}, "
+                    f"local shard metadata placement device: {local_device}"
                 )
 
-            if not local_shard_tensor.is_contiguous():
-                raise ValueError('Only torch.contiguous_format memory_format is currently supported')
+            if local_shard_tensor.dtype != local_shards_dtype:
+                raise ValueError(
+                    f"Local shards' tensor dtype property need to be the same! "
+                    f"Found one local shard tensor dtype: {local_shards_dtype}, "
+                    f"the other local shard tensor dtype: {local_shard_tensor.dtype}"
+                )
 
-            if shard_meta.shard_lengths != list(local_shard_tensor.size()):
+            if local_shard_tensor.requires_grad != local_shards_requires_grad:
                 raise ValueError(
-                    f'Local shard tensor is incompatible with local ShardMetadata! '
-                    f'local shard tensor size: {local_shard_tensor.size()}, '
-                    f'local ShardMetadata shard lengths: {shard_meta.shard_lengths}'
+                    f"Local shards' tensor requires_grad property need to be the same! "
+                    f"Found one local shard tensor requires_grad: {local_shards_requires_grad}, "
+                    f"the other local shard tensor requires_grad: {local_shard_tensor.requires_grad}"
                 )
 
-            if local_shard_tensor.is_pinned() != tensor_properties.pin_memory:
+            if local_shard_tensor.is_pinned() != local_shards_is_pinned:
                 raise ValueError(
-                    f'Local shard tensor pin_memory does not match with tensor_properties! '
-                    f'local shard tensor pin_memory: {local_shard_tensor.is_pinned()}, '
-                    f'tensor_properties pin_memory: {tensor_properties.pin_memory}'
+                    f"Local shards' tensor pin_memory property need to be the same! "
+                    f"Found one local shard tensor is_pinned(): {local_shards_is_pinned}, "
+                    f"the other local shard tensor is_pinned(): {local_shard_tensor.is_pinned()}"
                 )
 
-            if local_shard_tensor.device != local_device:
+        # build a "local" ShardedTensorMetadata with all local shards on this rank
+        global_size = torch.Size(check_tensor_size(overall_size))
```

Comment:
nit: 

s/check_tensor_size/validate_and_flatten since what is done here is not just "checking" :)